storage/storage.go

```diff
@@ -40,36 +42,88 @@ func (rr *ETLSource) NextTest() (string, []byte, error) {
 	metrics.WorkerState.WithLabelValues("read").Inc()
 	defer metrics.WorkerState.WithLabelValues("read").Dec()
 
-	h, err := rr.Next()
-	if err != nil {
-		return "", nil, err
-	}
-	if h.Typeflag != tar.TypeReg {
-		return h.Name, nil, nil
-	}
-	var data []byte
-	if strings.HasSuffix(strings.ToLower(h.Name), "gz") {
-		// TODO add unit test
-		zipReader, err := gzip.NewReader(rr)
+	// Try to get the next file.  We retry multiple times, because sometimes
+	// GCS stalls and produces stream errors.
+	// TODO - keep track of elapsed time instead of trials ??
+	trial := 0
+	delay := 5 * time.Millisecond
+	phase := "Next"
+	for {
```

Comment:
This loop is 80 lines and contains multiple `continue`s nested really deeply in it. I agree that it works, but would really like it to be simpler and clearer if possible.